Content.Server/StationEvents/Events/RandomSentienceRule.cs

```diff
@@ -2,19 +2,38 @@
 using Content.Server.Chat.Systems;
 using Content.Server.GameTicking.Rules.Components;
 using Content.Server.Ghost.Roles.Components;
+using Content.Server.Station.Components;
 using Content.Server.Station.Systems;
 using Content.Server.StationEvents.Components;
+using Robust.Shared.Utility;
 
 namespace Content.Server.StationEvents.Events;
 
 public sealed class RandomSentienceRule : StationEventSystem<RandomSentienceRuleComponent>
 {
+    [Dependency] private readonly StationSystem _stationSystem = default!;
+
     protected override void Started(EntityUid uid, RandomSentienceRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
     {
         HashSet<EntityUid> stationsToNotify = new();
 
+        var targetStation = _stationSystem.GetStations().FirstOrNull();
+
+        if (!TryComp(targetStation, out StationDataComponent? data))
+        {
+            Logger.Info("TargetStation not have StationDataComponent");
+            return;
+        }
+
         var mod = GetSeverityModifier();
-        var targetList = EntityQuery<SentienceTargetComponent>().ToList();
+        var targetList = EntityQuery<SentienceTargetComponent, TransformComponent>().ToList();
+
+
+        var grids = data.Grids.ToHashSet();
+        targetList.RemoveAll(
+            backupSpawnLoc =>
+                backupSpawnLoc.Item2.GridUid.HasValue && !grids.Contains(backupSpawnLoc.Item2.GridUid.Value));
+
```

Comment:
As above, also do not just copypaste the code onto 3 systems as it triples the work in maintaining it, add a method somewhere.